src/librustc_data_structures/dependency/state.rs

```diff
@@ -0,0 +1,157 @@
+// Copyright 2012-2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use fnv::{FnvHashSet, FnvHashMap};
+use graph::{Graph, NodeIndex};
+
+use super::DepNodeId;
+
+pub struct DepGraphState<ID: DepNodeId> {
+    graph: Graph<ID, ()>,
+    nodes: FnvHashMap<ID, NodeIndex>,
+    edges: FnvHashSet<(NodeIndex, NodeIndex)>,
+    open_nodes: Vec<OpenNode>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+enum OpenNode {
+    Node(NodeIndex),
+    Ignore,
+}
+
+impl<ID: DepNodeId> DepGraphState<ID> {
+    pub fn new() -> DepGraphState<ID> {
+        DepGraphState {
+            graph: Graph::new(),
+            nodes: FnvHashMap(),
+            edges: FnvHashSet(),
+            open_nodes: Vec::new()
+        }
+    }
+
+    /// Creates an entry for `node` in the graph.
+    fn make_node(&mut self, node: ID) -> NodeIndex {
+        if let Some(&i) = self.nodes.get(&node) {
+            return i;
+        }
+
+        let index = self.graph.add_node(node.clone());
+        self.nodes.insert(node, index);
+        index
+    }
+
+    /// Top of the stack of open nodes.
+    fn current_node(&self) -> Option<OpenNode> {
+        self.open_nodes.last().cloned()
+    }
+
+    /// All nodes reachable from `node`. In other words, things that
+    /// will have to be recomputed if `node` changes.
+    pub fn dependents(&self, node: ID) -> Vec<ID> {
+        match self.nodes.get(&node) {
+            None => vec![],
+            Some(&index) =>
+                self.graph.depth_traverse(index)
+                          .map(|dependent_node| self.graph.node_data(dependent_node).clone())
+                          .collect()
+        }
+    }
+
+    pub fn push_ignore(&mut self) {
+        self.open_nodes.push(OpenNode::Ignore);
+    }
+
+    pub fn pop_ignore(&mut self) {
+        let popped_node = self.open_nodes.pop().unwrap();
+        assert_eq!(popped_node, OpenNode::Ignore);
+    }
+
+    pub fn push_task(&mut self, key: ID) {
+        let top_node = self.current_node();
+
+        let new_node = self.make_node(key.clone());
+        self.open_nodes.push(OpenNode::Node(new_node));
+
+        // if we are in the midst of doing task T, then this new task
+        // N is a subtask of T, so add an edge N -> T.
```

Comment:
> The reason I'm asking is that this seems to be very subtle and it's not immediately obvious that it should be general behavior.

I think part of my assumption was that we should avoid nested tasks for the most part unless one task is a kind of "subtask" of the other -- either because it produces a result that the outer task needs, or because for the outer task to be considered "successful", the inner tasks must complete. In either case, a read edge is appropriate. 

In the code, this is generally true, particularly since `visit_all_items` removes "artificial" nesting due to lexical nesting (and the only cases that remain are when this nesting is kind of important).

Still, I could imagine another helper would be warranted.